### ISSUE TITLE
[BUG] On MSVC, _Complex is not a keyword

### ISSUE THREAD

**Issue by mattip** [cython/cython#5512]
### Describe the bug

When trying to use `complex.h` in NumPy in numpy/numpy#24085, we are seeing a failure around these lines https://github.com/cython/cython/blob/0d0cc9987f1c7e4d757d974be681e0288f0640e9/Cython/Utility/Complex.c#L85-L95

For C that expands to something like `typedef float _Complex __pyx_t_float_complex;`.

It turns out on MSVC `_Complex` is not a keyword rather a macro, so that line errors. [this page](https://learn.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=msvc-170) states (I could not find a link to the actual paragraph):
> MSVC doesn't support the _Complex keyword or native complex types. The Universal CRT <complex.h> uses implementation-specific macros to achieve the same effect. For more information, see [C complex math support](https://learn.microsoft.com/en-us/cpp/c-runtime-library/complex-math-support?view=msvc-170).

### Code to reproduce the behaviour:

```cython
# example code
```

I will try to fill this in later, there must be a failing test already in Cython...

### Expected behaviour

On MSVC, the typedef needs to be `typedef _FComplex __pyx_t_float_complex;`, see [this table](https://learn.microsoft.com/en-us/cpp/c-runtime-library/complex-math-support?view=msvc-170#types-used-in-complex-math)

### OS

Windows

### Python version

any

### Cython version

0.29.26, 3.0.0, master

**Comment by da-woods** [comment]
I think it's mostly the complex_numbers_*_T398 that test the various combinations of this. Not immediately sure why we don't catch this on some of the Windows builds but I'm sure it'll become clear

**Comment by da-woods** [comment]
I think most of the time `CYTHON_CCOMPLEX` doesn't get forced on by default. For it to be on by default we essentially require C11 + the relevant C11 feature check. Which I suspect MSVC doesn't have. So we likely don't have a c test where we force it on for MSVC (we likely have a C++ test, but C++ complex is better supported)

**Comment by mattip** [comment]
> I think it's mostly the `complex_numbers_*_T398` tests ...

> I think most of the time `CYTHON_CCOMPLEX` doesn't get forced on by default

Thanks. In NumPy, we started getting `CYTHON_CCOMPLEX=1` when the code has `#include <complex.h>`. There is [`complex_int_T446`](https://github.com/cython/cython/blob/0d0cc9987f1c7e4d757d974be681e0288f0640e9/tests/run/complex_int_T446.pyx#L7) which explicitly turns off `CYTHON_CCOMPLEX` for MSVC

**Comment by da-woods** [comment]
So I think my view is: we could special case it for MSVC as you suggest. It's definitely too late for 0.29.x.

If we could do it quickly then I *think* we could get it in 3.0. I'd be prepared to update the typedefs, but not actually change Cython to use `CYTHON_CCOMPLEX=1` by default (although if Numpy wants to do so, then that's your choice). I think doing that is safe because it shouldn't break any existing code.

I'll try to prepare a PR for this this weekend. I'll leave the final judgement on whether to include it to Stefan though.

-------------------------

In terms of Numpy's immediate use, I suspect you could use the fallback struct and just do a cast (they should have the same memory layout as all the compiler implementations) but that may not be pretty.

**Comment by mattip** [comment]
@da-wood: would it be fair to summarize the work done in #5517 as: it turns out the C support for complex in MSVC is very incomplete, and there does not seem to be a good way to get it to work without significant refactoring.

**Comment by da-woods** [comment]
I think you tagged someone with a similar (but different) name to me - hopefully they aren't too upset by having to think about MSVC complex numbers :).

The big thing that's missing from the MSVC support is the basic arithmetic operators.  While we have fallback implementations of those in Cython (for the case where we use our own complex type), I can't see a way to use them quickly and simply.  I'm sure it's possible but it's gone a bit beyond "a quick change that I feel comfortable sneaking into the 3.0 release candidate".

I'll have a bit more of a think about it and if I come up with anything I'll give it another go.

I should also add, I think we use non-standard gcc extensions too, so it's not 100% microsoft's fault.

**Comment by lysnikolaou** [comment]
Thanks a lot for taking a stab at this. It's challenging indeed.

I think the proper fix here is not to check for `_Complex_I` under MSVC. I'm referring to the code snippet over here: https://github.com/cython/cython/blob/52cba8252b3527f0ca998378ba9bd2b8d5d9ee64/Cython/Utility/Complex.c#L4-L13

This appears to be wrong, cause MSVC does not support the `_Complex` keyword, but it *does* define the `_Complex_I` macro. How would you feel about falling back to the Cython structure under MSVC, i.e. change the above snippet to this:

```c
#if !defined(CYTHON_CCOMPLEX) 
  #if defined(__cplusplus) 
    #define CYTHON_CCOMPLEX 1 
  #elif (defined(_Complex_I) && !defined(_MSC_VER)) || ((defined (__STDC_VERSION__) && __STDC_VERSION__ >= 201112L) && !defined(__STDC_NO_COMPLEX__)) 
    // <complex.h> should exist since C99, but only C11 defines a test to detect it 
    #define CYTHON_CCOMPLEX 1 
  #else 
    #define CYTHON_CCOMPLEX 0 
  #endif 
#endif 
```

Keep in mind that I don't have any experience with Cython and its handling of complex number, so I might be misunderstanding something here.

**Comment by da-woods** [comment]
Ah I see... Maybe I've misunderstood the real problem.

So you aren't intentionally defining `CYTHON_CCOMPLEX` yourself, but you're finding it's getting used because `defined(_Complex_I)` is true?

We can definitely fix that.

(In the shorter term Numpy can always define `CYTHON_CCOMPLEX 0` itself as a compile-time define, to override Cython's autodetection)

**Comment by lysnikolaou** [comment]
> So you aren't intentionally defining CYTHON_CCOMPLEX yourself, but you're finding it's getting used because defined(_Complex_I) is true?

That's correct.

**Comment by da-woods** [comment]
In that case, your suggested change looks right. Feel free to submit a PR (although I'll do it myself if one doesn't appear fairly shortly)

**Comment by lysnikolaou** [comment]
I can submit a PR early tomorrow if that works for you.

**Comment by scoder** [comment]
I pushed the proposed change in https://github.com/cython/cython/commit/d8bb216c261d1911a0725ea9364cb837e692416b

**Comment by scoder** [comment]
I also added a minimal change to 0.29.x in https://github.com/cython/cython/commit/436f8a32ea3a1b0411695db1b974c60434042b5b because I wasn't sure if the "stdc" check in the master branch is complete by itself. Either way, this fix will (hopefully) be released in 3.0 first.
